---
layout: page
title: People
permalink: /people/
---

<div class="center">

<div class="item">
    <img class="rcornersimage" src="/assets/images/sebastian.jpg">
    <h5><div class="tooltip"><p class="ttext">Sebastian Höhna</p>
        <span class="tooltiptext">
            I’m a computational biologist with focus on Bayesian phylogenetic inference. Natural selection by means of adaptation and genetic inheritance are key principles in evolutionary biology. Evolutionary relationships are commonly depicted by phylogenetic trees. Using phylogenetic methods we can learn about the evolutionary history of species and the processes that have contributed to present diversity. I’m developing statistical and computational methods to infer phylogenies from molecular sequence data. These methods additionally identify periods of adaptive genetic evolution at lineage or genes. Furthermore, I develop mathematical models to study macroevolutionary patterns, such as, diversification rate variation over time and among lineages and episodes of global mass extinctions.
        </span>
    </div></h5>
        <p class="title"><em>Group Leader</em></p>
        <a href="https://github.com/hoehna" target="_blank"><img src="/assets/icons/mark-github.svg" width="25px"></a>
        <a href="https://twitter.com/sebastianhoehna?lang=en" target="_blank"><img src="/assets/icons/twitter.png" width="25px"></a>
        <a href="mailto:hoehna@lmu.de"><img src="/assets/icons/mail.svg" width="25px"></a>

    <!-- Show tooltip text when on mobile -->
    <div class="blurb">
        I’m a computational biologist with focus on Bayesian phylogenetic inference. Natural selection by means of adaptation and genetic inheritance are key principles in evolutionary biology. Evolutionary relationships are commonly depicted by phylogenetic trees. Using phylogenetic methods we can learn about the evolutionary history of species and the processes that have contributed to present diversity. I’m developing statistical and computational methods to infer phylogenies from molecular sequence data. These methods additionally identify periods of adaptive genetic evolution at lineage or genes. Furthermore, I develop mathematical models to study macroevolutionary patterns, such as, diversification rate variation over time and among lineages and episodes of global mass extinctions.
    </div>
</div>

<div class="item">
    <img class="rcornersimage" src="/assets/images/allison.jpg">
    <h5><div class="tooltip"><p class="ttext">Allison Hsiang</p>
        <span class="tooltiptext">
            My research uses statistical modeling, Bayesian inference, Big Data, and machine learning to understand morphological evolution and macroevolutionary patterns and processes. I am particularly interested in bringing a ‘next-generation’ approach (e.g., high-throughput automated data generation) to studying fossils and morphology in a phylogenetic context. I am also interested in the interplay between morphology and genetics, including methodological considerations during phylogenetic inference.
        </span>
    </div></h5>
        <p class="title"><em>Postdoctoral Researcher</em></p>
        <a href="https://github.com/ahsiang" target="_blank"><img src="/assets/icons/mark-github.svg" width="25px"></a>
        <a href="https://ahsiang.github.io/" target="_blank"><img src="/assets/icons/globe.svg" width="25px"></a>
        <a href="mailto:allison.hsiang@lmu.de"><img src="/assets/icons/mail.svg" width="25px"></a>

    <!-- Show tooltip text when on mobile -->
    <div class="blurb">
        My research uses statistical modeling, Bayesian inference, Big Data, and machine learning to understand morphological evolution and macroevolutionary patterns and processes. I am particularly interested in bringing a ‘next-generation’ approach (e.g., high-throughput automated data generation) to studying fossils and morphology in a phylogenetic context. I am also interested in the interplay between morphology and genetics, including methodological considerations during phylogenetic inference.
    </div>

</div>

<div class="item">
    <img class="rcornersimage" src="/assets/images/luiza.jpg">
    <h5><div class="tooltip"><p class="ttext">Luiza Fabreti</p>
        <span class="tooltiptext">
            I'm interested in computational and evolutionary biology. Specially, improving the way we model evolution. My research involves Bayesian phylogenetic inference and underlying topics such as, model selection, testing the adequacy of the most commonly used substitution models and assessing convergence of the Markov Chain Monte Carlo.
        </span>
    </div></h5>
        <p class="title"><em>Graduate Student</em></p>
        <a href="https://github.com/lfabreti" target="_blank"><img src="/assets/icons/mark-github.svg" width="25px"></a>
        <a href="mailto:l.fabreti@lmu.de"><img src="/assets/icons/mail.svg" width="25px"></a>

    <!-- Show tooltip text when on mobile -->
    <div class="blurb">
        I'm interested in computational and evolutionary biology. Specially, improving the way we model evolution. My research involves Bayesian phylogenetic inference and underlying topics such as, model selection, testing the adequacy of the most commonly used substitution models and assessing convergence of the Markov Chain Monte Carlo.
    </div>
</div>

<div class="item">
    <img class="rcornersimage" src="/assets/images/ronja.jpg">
    <h5><div class="tooltip"><p class="ttext">Ronja Billenstein</p>
        <span class="tooltiptext">
            I’m a biologist working on statistical methods for inferring the evolutionary history of closely related populations. These methods are based on coalescent theory and a mutation process. I am using simulated as well as empirical data to test and to validate these methods, hoping to provide an accurate, powerful and computationally feasible approach.
        </span>
    </div></h5>
        <p class="title"><em>Graduate Student</em></p>
        <a href="mailto:r.billenstein@lrz.uni-muenchen.de"><img src="/assets/icons/mail.svg" width="25px"></a>

    <!-- Show tooltip text when on mobile -->
    <div class="blurb">
        I’m a biologist working on statistical methods for inferring the evolutionary history of closely related populations. These methods are based on coalescent theory and a mutation process. I am using simulated as well as empirical data to test and to validate these methods, hoping to provide an accurate, powerful and computationally feasible approach.
    </div>
</div>

<div class="item">
    <img class="rcornersimage" src="/assets/images/killian.jpg">
    <h5><div class="tooltip"><p class="ttext">Killian Smith</p>
        <span class="tooltiptext">
I am a computer scientist working with Bayesian phylogenetics and substitution models, and am
interested in applying concepts and theory from CS in this domain. In my research I will be
exploring covarion models, mixture models, and Markov modulated continuous Markov chains (MMCMC),
and their effects on phylogenetic tree construction. Some of my other research interests include
type theory, functional programming, de novo sequencing, multiple sequence alignment, genome
assembly, and machine learning.
        </span>
    </div></h5>
        <p class="title"><em>Graduate Student</em></p>
        <a href="mailto:k.smith@lrz.uni-muenchen.de"><img src="/assets/icons/mail.svg" width="25px"></a>

    <!-- Show tooltip text when on mobile -->
    <div class="blurb">
I am a computer scientist working with Bayesian phylogenetics and substitution models, and am
interested in applying concepts and theory from CS in this domain. In my research I will be
exploring covarion models, mixture models, and Markov modulated continuous Markov chains (MMCMC),
and their effects on phylogenetic tree construction. Some of my other research interests include
type theory, functional programming, de novo sequencing, multiple sequence alignment, genome
assembly, and machine learning.
    </div>
</div>

</div>
