---
layout: page
title: Publications
permalink: /publications/
---

<h3>2019</h3>
- Freyman WA and <strong>S Höhna</strong>: Stochastic Character Mapping of State-Dependent Diversification Reveals the Tempo of Evolutionary Decline in Self-Compatible Onagraceae Lineages, Systematic Biology, 2019, 68 (3), 505--519, [link](https://academic.oup.com/sysbio/article-abstract/68/3/505/5200730)
- Silvestro D, MF Tejedor, ML Serrano-Serrano, O Loiseau, V Rossier, J Rolland, A Zizka, <strong>S Höhna</strong>, A Antonelli and N Salamin: Early Arrival and Climatically-Linked Geographic Expansion of New World Monkeys from Tiny African Ancestors, Systematic Biology, 2019, 68 (1), 78--92, [link](https://academic.oup.com/sysbio/article/68/1/78/5040681)

<h3>2018</h3>
- Condamine FL, J Rolland J, <strong>S Höhna</strong>, FAH Sperling and I Sanmartin I: Testing the Role of the Red Queen and Court Jester as Drivers of the Macroevolution of Apollo Butterflies, Systematic Biology, 2018, 67 (6), 940--964, [link](https://academic.oup.com/sysbio/article-abstract/67/6/940/4847851)
- Martin CH and <strong>S Höhna</strong>: New evidence for the recent divergence of Devil's Hole pupfish and the plausibility of elevated mutation rates in endangered taxa, Molecular Ecology, 2018, 27 (4), 831--838, [link](https://onlinelibrary.wiley.com/doi/full/10.1111/mec.14404)
- <strong>Höhna S</strong>, LM Coghill, G Mount, R Thomson and JM Brown: P<sup>3</sup>: Phylogenetic Posterior Prediction in RevBayes, Molecular Biology and Evolution, 2018, 35 (4), 1028--1034, [link](https://academic.oup.com/mbe/article/35/4/1028/4616601)
- Freyman WA and <strong>Höhna S</strong>: Cladogenetic and Anagenetic Models of Chromosome Number Evolution: a Bayesian Model Averaging Approach, Systematic Biology, 2018, 67 (2), 195--215, [link](https://academic.oup.com/sysbio/article/67/2/195/4037179)


<h3>2017</h3>
- <strong>Höhna S</strong>, MJ Landis and TA Heath: Phylogenetic Inference using RevBayes. Current Protocols in Bioinformatics, 2017, 57:6.16.1-6.16.34, [link](http://onlinelibrary.wiley.com/doi/10.1002/cpbi.22/full)
- Martin CH, <strong>S Höhna</strong>, JE Crawford, BJ Turner, EJ Richards and LH Simons: The complex effects of demographic history on the estimation of substitution rate: concatenated analysis results in no more than 2-fold overestimation, Proceedings of the Royal Society B, 2017, 284: 20170537, [link](https://royalsocietypublishing.org/doi/full/10.1098/rspb.2017.0537)

<h3>2016</h3>
- Moore BR, <strong>S Höhna</strong>, MR May, B Rannala and JP Huelsenbeck: Critically evaluating the theory and performance of Bayesian analysis of macroevolutionary mixtures. Proceedings of the National Academy of Sciences, 2016, 113 (34), 9569-9574, [link](https://www.pnas.org/content/113/34/9569.short)
- <strong>Höhna S</strong>, MJ Landis, TA Heath, B Boussau, BR Moore, N Lartillot, JP Huelsenbeck and F Ronquist: RevBayes: Bayesian Phylogenetic Inference Using Graphical Models and an Interactive Model-Speci cation Language, Systematic Biology, 2016, 65 (4), 726-736, [link](https://academic.oup.com/sysbio/article/65/4/726/1753608/RevBayes-Bayesian-Phylogenetic-Inference-Using)
- Conroy CR, JL Patton, M Lim, M Phuong, B Parmenter and <strong>S Höhna</strong>: Following the rivers: historical reconstruction of California voles (Microtus californicus, Muridae, Rodentia) in the deserts of eastern California, Biological Journal of the Linnean Society, 2016, 119 (1), 80–98, [link](http://onlinelibrary.wiley.com/doi/10.1111/bij.12808/full)
- <strong>Höhna S</strong>, MR May and BR Moore: TESS: an R package for efficiently simulating phylogenetic trees and performing Bayesian inference of lineage diversification rates, Bioinformatics, 2016, 32 (5): 789-791, [link](https://academic.oup.com/bioinformatics/article/32/5/789/1744433/TESS-an-R-package-for-efficiently-simulating)
